File: miruzo-py/app/services/settings/timezone.py
from tzlocal import get_localzone

from app.persist.settings.protocol import SettingsRepository

_TIMEZONE_SETTINGS_KEY = 'timezone'


def _get_default_location() -> str:
	localzone = get_localzone()
	return localzone.key


class TimezoneProvider:
	def __init__(self, repository: SettingsRepository) -> None:
		self._repository = repository
		self._location = _get_default_location()

	@property
	def location(self) -> str:
		return self._location

	def set_location(self, value: str) -> None:
		self._location = value

	def ensure_settings(self, initial_location: str | None) -> bool:
		location = self._repository.get(_TIMEZONE_SETTINGS_KEY)
		if location is None:
			if initial_location is not None:
				self._location = initial_location
			self._repository.insert(_TIMEZONE_SETTINGS_KEY, self._location)
			return True

		self._location = location
		return False

File: miruzo/internal/service/settings/timezone.go
package settings

import (
	"context"
	"errors"
	"log"
	"time"

	"github.com/mntone/miruzo-core/miruzo/internal/adapter/timezone"
	"github.com/mntone/miruzo-core/miruzo/internal/persist"
	"github.com/samber/mo"
)

const timezoneSettingsKey = "timezone"

type TimezoneProvider struct {
	repository persist.SettingsRepository
	resolver   timezone.TimezoneResolver
	location   mo.Option[string]
}

func NewTimezoneProvider(
	repository persist.SettingsRepository,
	resolver timezone.TimezoneResolver,
) *TimezoneProvider {
	return &TimezoneProvider{
		repository: repository,
		resolver:   resolver,
	}
}

func (prov *TimezoneProvider) Location() string {
	value, present := prov.location.Get()
	if !present {
		log.Fatalf("settings: timezone is not initialized; call EnsureSettings first")
	}

	return value
}

func (prov *TimezoneProvider) SetLocation(location string) {
	prov.location = mo.Some(location)
}

func (prov *TimezoneProvider) saveLocation(
	ctx context.Context,
	location string,
) {
	err := prov.repository.UpdateValue(ctx, timezoneSettingsKey, location)
	if err != nil {
		log.Printf(
			"settings: failed to save timezone: key=%s value=%s",
			timezoneSettingsKey,
			location,
		)
	}
}

func (prov *TimezoneProvider) EnsureSettings(
	ctx context.Context,
	initialLocation *string,
) {
	// 1. Use the saved location from the settings table.
	appLocation, err := prov.repository.GetValue(ctx, timezoneSettingsKey)
	if err == nil {
		prov.location = mo.Some(appLocation)
		return
	}
	if !errors.Is(err, persist.ErrNotFound) {
		log.Fatalf("settings: failed to load timezone: %v", err)
	}

	// 2. If missing, use/save the initial location.
	if initialLocation != nil {
		prov.saveLocation(ctx, *initialLocation)
		prov.location = mo.Some(*initialLocation)
		return
	}

	// 3. If still missing, use/save the host's location.
	systemLocation, present := prov.resolver.GetLocation().Get()
	if present {
		prov.saveLocation(ctx, systemLocation)
		prov.location = mo.Some(systemLocation)
		return
	}

	// 4. If still missing, use/save UTC.
	utcLocation := time.UTC.String()
	prov.saveLocation(ctx, utcLocation)
	prov.location = mo.Some(utcLocation)
}
